design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+
+#### POST
+
+HTTP `POST` should be used to create new resources. It is a non-safe and non-idempotent operation and clients should be aware of that.
+
+Successful `POST` requests should always result in a `201` status code and empty response. The response should also contain a reference
+to the resource that was created:
+
+```
+HTTP/1.1 201 Created
+Content-Length: 0
+Location: http://mykeycloak.com/apis/users/{version}/1234
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|201|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+#### PUT
+
+HTTP `PUT` should be used to replace a resource state.
+
+Successful `PUT` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+#### DELETE
+
+HTTP `DELETE` should be used to remove a resource.
+
+Successful `DELETE` requests should always result in a `204` status code and empty response.
+
+```
+HTTP/1.1 204 No Content
+Content-Length: 0
+```
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `POST` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|204|
+|400|
+|401|
+|403|
+|404|
+|409|
+|500|
+
+## Practices
+
+### Error Handling
+
+A good error handling mechanism helps to communicate to clients the reasons why a request could not be fulfilled and eventually
+recover from this errors.
+
+When returning errors, developers should use a common [HTTP Status Code](#httpstatuscodes) and include a body using JSON as the format with the information
+why a request failed.
+
+The error message should have the following format:
+
+|Claim|Description|Required|
+|---|---|---|
+|error|The error code|Yes|
+|error_description|A text providing more details for clients about the error and eventually how to recover from it|No|
+
+Here is an example of a error response:
+
+```json
+HTTP/1.1 400 Bad Request
+{
+    "error": "invalid_request",
+    "error_description": "The field 'email' is mandatory"
+}
+```
+
+#### Error Codes
+
+Error codes can be any string. They should be lowercase and use `_` when the code has multiple words. However, developers should 
+consider existing error codes before creating new ones.
+
+When processing error responses, developers should look the error codes being used by other APIs and see if they can be reused.
+
+### Filtering
+
+TODO
+
+### Pagination
+
+TODO
+
```

Comment:
Should we also consider implementation of some sort of abuse rate ([similarly to GitHub](https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits))?

Reply:
@abstractj, did some research about pagination. I did not have the chance to check his proposal at this regard. I'm open to discuss all the practices we find useful to base our own.

Reply:
This comment is not exactly about pagination. :) It's just about the abuse rate in the context of whole REST API. (I should've put the comment somewhere else so it wouldn't be confusing.)

Reply:
Ahh sorry. I did associate it with the pagination practice.

Your suggestion seems to be more related with recommendations, what if we provide a third top-level concept to the guideline specific for that? We could probably use this part of the guideline to include any recommendation we find necessary for consuming our APIs.

Reply:
I believe rate limiting could be part of the Security section.

Reply:
>I believe rate limiting could be part of the Security section.
+1